python/lsst/cp/pipe/utilsEfd.py

```diff
@@ -202,11 +206,67 @@ def selectTimeSeries(self, topicName, fields=[], startDate=None, endDate=None):
         table = Table(rows=series["values"], names=series["columns"], dtype=tableDtype)
         table["time"] = Time(table["time"], scale="utc")
         table.sort("time")
+        if 'private_sndStamp' in table.columns:
+            table['private_sndStamp'] = Time(table['private_sndStamp'], format='unix_tai')
+            table.sort("private_sndStamp")
 
         return table
 
+    def searchResults(self, data, dateStr, scale='tai'):
+        """Determine the entry for a specific date.
+
+        Parameters
+        ----------
+        data : `astropy.table.Table`
+            The table of results from the EFD.
+        dateStr : `str`
+            The date to look up in the status for.
+        scale : `str`, optional
+            Time scale to use.  Default is 'tai'.
+
+        Returns
+        -------
+        result = `astropy.table.Row`
+            The row of the data table corresponding to ``dateStr``.
+        """
+        dateValue = Time(dateStr, format='isot', scale=scale)
+        # Table is now sorted on "time", which is in UTC.
```

Comment:
I thought it was TAI?

Reply:
`time` is UTC, `private_sndStamp` should be TAI.  Except for the electrometers.  Which are UTC and so needed this scale addition.  I've chosen the path that was most likely to get things working, and we can revisit this if we can get all of the devices using a consistent time.

Reply:
The default and the comments don’t make sense to me, but these time issues are always hard. It says utc but above the default is tai…

Reply:
tl;dr: Everything is in TAI now, and it appears to be consistent.

I hope I never end up in another UTC/TAI rabbit hole.  That said:

1. The filter scan task uses of the cp_pipe EFD client explicitly cast `visitInfo.date` to `DateTime.TAI` and uses that for the query.
2. The monochromator scan task uses the `DATE-BEG` header keyword, as the fiber spectrograph doesn't make a valid visitInfo yet, but that is defined as TAI in the headers.
3. The column `time` is a string representation of UTC, whether accessed via electrometer `intensity` or `logMessage` or via the monochromator `wavelength` series.  I've messed up the numbers here.
6. The column `private_sndStamp` is an float `unix_tai` value which is indeed in TAI.
7. `blah = Time(data['private_sndStamp'], format='unix_tai')` gives an `astropy.Time` version of the TAI value, which can be converted to a useful string representation with `blah.isot` and `blah.to_value('isot')`.
8. `blah.utc.isot` and `blah.utc.to_value('isot')` yield UTC string representations, as does `Time(data['private_sndStamp'], scale='utc', format='unix_tai').isot`.

Point 3 is forced, as the raw value received from the EFD is a string like `2022-10-05T14:47:49.764833Z`, which, due to the trailing `Z` must have `scale='utc'` or astropy complains.  Because that is what the EFD values are, we need to cast any start/end dates to UTC, which we do in `selectTimeSeries` when doing the query.  However, we rewrite the `time` column, but leave it in UTC.  I've now changed the code so that `time` is now TAI, making Point 3 no longer true.  Now the only method that has any conversions is `selectTimeSeries`, which now expects TAI start/end dates, will do the cast to UTC for the query, and converts the returned values in `time` to TAI, so `time` should now match `private_sndStamp`.  We've already had a `sort` on `time`, so this means the table is now monotonic, and universally in TAI.

If everything is in TAI, then `searchResults` no longer needs a `scale` argument, so I've removed that.  We actually do the search there against `private_sndStamp`, so this should have been in TAI all along.  Everything user-facing is now TAI.  Everything is TAI.

Running the tests, `test_electrometer` now fails, with `-1.5232e-07 != -1.8932e-07` where it's failing on the first lookup.  This shows that I have to ensure that the search string is converted with `scale='tai'`.  Which now causes issues because the test dates were poorly chosen.  Fixing that by ensuring the first date is included in the data set.

Now the `test_electrometer_alternate` is failing, but I can see by eye that's because the test dates are UTC, and not TAI as I've now asserted everything is in.  Fixing that results in all tests passing, so I'm going to commit and push.